# Activity 5.1 Text Similarity

#### Objective(s):

This activity aims to introduce how to implement text similarity

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to remove special characters from text
* Demonstrate how to measure text simiarity using distance
* Demonstrate how to find text similarity using Count Vector and TF-IDF Vectorizer


#### Resources:
* Jupyter Notebook
* songdata.csv

#### Procedures
Load the necessary libraries

In [2]:
import nltk
import pandas as pd

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
data = pd.read_csv('/content/drive/MyDrive/SCHOOLFILES/CPE32/CPE313/hoa8.1/songdata.csv')
data.head()

,artist,song,link,text
0,ABBA,Ahe's My Kind Of Girl,/a/abba/ahes+my+kind+of+girl_20598417.html,"Look at her face, it's a wonderful face \r\nA..."
1,ABBA,"Andante, Andante",/a/abba/andante+andante_20002708.html,"Take it easy with me, please \r\nTouch me gen..."
2,ABBA,As Good As New,/a/abba/as+good+as+new_20003033.html,I'll never know why I had to go \r\nWhy I had...
3,ABBA,Bang,/a/abba/bang_20598415.html,Making somebody happy is a question of give an...
4,ABBA,Bang-A-Boomerang,/a/abba/bang+a+boomerang_20002668.html,Making somebody happy is a question of give an...


Find the lyrics by the artist The Beatles

In [6]:
beatles = data[data.artist=='The Beatles']
beatles.head()

,artist,song,link,text
1198,The Beatles,A Shot Of Rhythm And Blues,/b/beatles/a+shot+of+rhythm+blues_20014867.html,"Well, if your hands start a-clappin' \r\nAnd ..."
1199,The Beatles,Across The Universe,/b/beatles/across+the+universe_10026507.html,Words are flowing out like \r\nEndless rain i...
1200,The Beatles,All I've Got To Do,/b/beatles/all+ive+got+to+do_10026646.html,"Whenever I want you around, yeah \r\nAll I go..."
1201,The Beatles,And I Love Her,/b/beatles/and+i+love+her_10026463.html,I give her all my love \r\nThat's all I do \...
1202,The Beatles,And Your Bird Can Sing,/b/beatles/and+your+bird+can+sing_10026364.html,You tell me that you've got everything you wan...


Check the number of Beatles songs in the data set


In [7]:
data.artist.value_counts().loc[['The Beatles']]

,count
artist,
The Beatles,178


Take a look at the first song's lyrics

In [8]:
beatles.iloc[0,3]

"Well, if your hands start a-clappin'  \r\nAnd your fingers start a-poppin'  \r\nAnd your feet start a-movin' around  \r\nAnd if you start to swing and sway  \r\n  \r\nWhen the band starts to play  \r\nA real cool way out sound  \r\nAnd if you get to can't help it and you can't sit down  \r\nYou feel like you gotta move around  \r\n  \r\nYou get a shot of rhythm and blues.  \r\nWith just a little rock and roll on the side  \r\nJust for good measure.  \r\nGet a pair of dancin' shoes  \r\n  \r\nWell, with your lover by your side  \r\nDon't you know you're gonna have a rockin' time, see'mon!  \r\nDon't you worry 'bout a thing  \r\nIf you start to dance and sing  \r\n  \r\nAnd chills come up on you  \r\nAnd if the rhythm finally gets you and the beat gets you too  \r\nWell, here's something for you to do  \r\n  \r\nGet a shot of rhythm and blues  \r\nWith just a little rock and roll on the side  \r\nJust for good measure  \r\nGet a pair of dancin' shoes  \r\n  \r\nWell, with your lover by 

Remove characters from the lyrics

In [9]:
import re

newline = lambda x: re.sub('\n', ' ', x) # remove \n
alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words

corpus = beatles.text.map(newline).map(alphanumeric)
corpus.head()

<>:4: SyntaxWarning: invalid escape sequence '\w'
<>:4: SyntaxWarning: invalid escape sequence '\w'
/tmp/ipykernel_11243/2279837142.py:4: SyntaxWarning: invalid escape sequence '\w'
  alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words


,text
1198,"Well, if your hands start a-clappin' \r And y..."
1199,Words are flowing out like \r Endless rain in...
1200,"Whenever I want you around, yeah \r All I got..."
1201,I give her all my love \r That's all I do \r...
1202,You tell me that you've got everything you wan...


Create a document term matrix using Count Vectorizer with the stop words turned on to English

In [11]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(stop_words="english")
X = cv.fit_transform(corpus).toarray()

dt = pd.DataFrame(X, columns=cv.get_feature_names_out()).set_index(beatles.song)
dt.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Across The Universe,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
All I've Got To Do,0,0,0,0,0,0,0,0,0,0,...,2,0,0,0,0,0,0,0,0,0
And I Love Her,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
And Your Bird Can Sing,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [35]:
dt.shape

(178, 2026)

Count vecotrizer converts text into numbers by building a vocabulary of every unique word across all 178 songs. Then, for each song, it counts how many time each vocabulary word appears. The result is a 178 x 2026 matrix, which is something that can be used for classifiers and other tasks.

Set display options so that the lyrics aren't cut off

In [13]:
pd.set_option('display.max_colwidth', None)

In [14]:
beatles[beatles.song=='Imagine'].text

,text
24783,Imagine there's no heaven \r\nIt's easy if you try \r\nNo hell below us \r\nAbove us only sky \r\nImagine all the people \r\nLiving for today... \r\n \r\nImagine there's no countries \r\nIt isn't hard to do \r\nNothing to kill or die for \r\nAnd no religion too \r\nImagine all the people \r\nLiving life in peace... \r\n \r\nYou may say I'm a dreamer \r\nBut I'm not the only one \r\nI hope someday you'll join us \r\nAnd the world will be as one \r\n \r\nImagine no possessions \r\nI wonder if you can \r\nNo need for greed or hunger \r\nA brotherhood of man \r\nImagine all the people \r\nSharing all the world... \r\n \r\nYou may say I'm a dreamer \r\nBut I'm not the only one \r\nI hope someday you'll join us \r\nAnd the world will live as one\r\n\r\n


Imagine lyrics in Count Vectorizer form


In [36]:
imagine = list(dt.loc['Imagine'])
imagine[:20]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Define the cosine similarity calculation

In [16]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

Cosine similarity measures the angle between two vectors. If two songs use very similar words, their vectors point in the same direction and the score is close to 1.0. If they share no words, the score is 0

Calculate all similarities and sort by the most similar

In [17]:
similarity = [cosine(imagine, song) for song in X]
sorted(list(zip(similarity, beatles.song)), reverse=True)[1:6]

[(np.float64(0.22757944185316942), "I'll Cry Instead"),
 (np.float64(0.1738270040532526), 'In My Life'),
 (np.float64(0.16574838603294895), 'Eleanor Rigby'),
 (np.float64(0.1490280831849844), 'In Spite Of All The Danger'),
 (np.float64(0.14592261028699943), 'All My Loving')]

Create the TF-IDF Vectorizer

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer

cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer

cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


TF-IDF works the same as count vecotrizer. However, instead of frequency count, it assigns a weight to each word. Words that appear a lot in one song but rarely in other songs get a high weight (these are distinctive)
Words that appear in many songs get a low weight (these are common and not helpful for comparison)

Calculate all similarities and sort by the most similar using the TF-IDF data

In [22]:
similarity_tfidf = [cosine(imagine, song) for song in X_tfidf]
sorted(list(zip(similarity_tfidf, beatles.song)), reverse=True)[1:6]

[(np.float64(0.19266277925834221), "I'll Cry Instead"),
 (np.float64(0.18004423848592196), "I'll Get You"),
 (np.float64(0.1551371744116933), 'Eleanor Rigby'),
 (np.float64(0.15054395512050195), 'In My Life'),
 (np.float64(0.12306875209548622), 'Love Of The Loved')]

This finds the top 5 songs most lyrically similar to "Imagine", using TF-IDF weights.

Compare the lyrics of Imagine and I'll Cry Instead

In [23]:
# Imagine lyrics
beatles[beatles.song=='Imagine'].text

,text
24783,Imagine there's no heaven \r\nIt's easy if you try \r\nNo hell below us \r\nAbove us only sky \r\nImagine all the people \r\nLiving for today... \r\n \r\nImagine there's no countries \r\nIt isn't hard to do \r\nNothing to kill or die for \r\nAnd no religion too \r\nImagine all the people \r\nLiving life in peace... \r\n \r\nYou may say I'm a dreamer \r\nBut I'm not the only one \r\nI hope someday you'll join us \r\nAnd the world will be as one \r\n \r\nImagine no possessions \r\nI wonder if you can \r\nNo need for greed or hunger \r\nA brotherhood of man \r\nImagine all the people \r\nSharing all the world... \r\n \r\nYou may say I'm a dreamer \r\nBut I'm not the only one \r\nI hope someday you'll join us \r\nAnd the world will live as one\r\n\r\n


In [24]:
beatles[beatles.song=='I\'ll Cry Instead'].text

,text
24774,"I've got every reason on earth to be mad \r\n'Cause I just lost the only girl I had \r\nIf I could get my way \r\nI'd get myself locked up today \r\nBut I can't, so I'll cry instead \r\n \r\nI've got a chip on my shoulder that's bigger that my feet \r\nI can't talk to people that I meet \r\nIf I could see you now \r\nI'd try to make you sad somehow \r\nBut I can't, so I'll cry instead \r\n \r\nDon't want to cry when there's people there \r\nI get shy when they start to stare \r\nI'm gonna hide myself away \r\nBut I'll come back again someday \r\n \r\nAnd when you do you'd better hide all the girls \r\nI'm gonna break their hearts all round the world \r\nYes, I'm gonna break them in two \r\nAnd show you what your lovin' man can do \r\nUntil then I'll cry instead\r\n\r\n"


Compare the two lyrics based on the output

*The lyrics of the two songs are largely different.*

Find the two Beatles songs that are the most similar
* Using Count Vectorizer
* Using TF-IDF Vectorizer

Calculate the cosine similarity between all combinations of documents

In [25]:

from itertools import combinations

pairs = list(combinations(beatles.song.index, 2)) # all song index combos
pairs_0 = list(combinations(range(len(beatles)), 2)) # all index combos starting with (0,1)
song_pairs = [(beatles.song[a_index], beatles.song[b_index]) for (a_index, b_index) in pairs]
song_pairs[:5]

[('A Shot Of Rhythm And Blues', 'Across The Universe'),
 ('A Shot Of Rhythm And Blues', "All I've Got To Do"),
 ('A Shot Of Rhythm And Blues', 'And I Love Her'),
 ('A Shot Of Rhythm And Blues', 'And Your Bird Can Sing'),
 ('A Shot Of Rhythm And Blues', 'Another Girl')]

 Calculate the cosine similarity of the Count Vectorizer document-term matrix

In [26]:

results = [cosine(X[a_index], X[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8548876342047523), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.8359173188630292),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.8058125835571367), ('And I Love Her', 'Love Me Do')),
 (np.float64(0.7930110315846619), ('If I Fell', 'Love Me Do')),
 (np.float64(0.7478391382739419), ('And I Love Her', 'All You Need Is Love'))]

Calculate the cosine similarity of the TF-IDF Vectorizer document-term matrix

In [27]:

results = [cosine(X_tfidf[a_index], X_tfidf[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8717047783261523),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.6599353568343105), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.6502842517883831), ('Have A Banana!', 'Crinsk Dee Night')),
 (np.float64(0.6232797327325461), ("Don't Let Me Down", 'Let It Be')),
 (np.float64(0.5275364336322863), ('If I Fell', 'Love Me Do'))]

In [28]:

beatles[beatles.song=='All You Need Is Love'].text

,text
24695,"Love, love, love, love, love, love, love, love, love. \r\n \r\nThere's nothing you can do that can't be done. \r\nNothing you can sing that can't be sung. \r\nNothing you can say, but you can learn \r\nHow to play the game \r\nIt's easy. \r\nNothing you can make that can't be made. \r\nNo one you can save that can't be saved. \r\nNothing you can do, but you can learn \r\nHow to be you in time \r\nIt's easy. \r\n \r\nAll you need is love, all you need is love, \r\nAll you need is love, love. Love is all you need. \r\nLove, love, love, love, love, love, love, love, love. \r\nAll you need is love, all you need is love, \r\nAll you need is love, love. Love is all you need. \r\n \r\nThere's nothing you can know that isn't known. \r\nNothing you can see that isn't shown. \r\nThere's nowhere you can be that isn't where \r\nYou're meant to be \r\nIt's easy. \r\n \r\nAll you need is love, all you need is love, \r\nAll you need is love, love. Love is all you need. \r\nAll you need is love. (All together now). \r\nAll you need is love. (Everybody). \r\nAll you need is love, love. Love is all you need. \r\nLove is all you need. \r\nLove is all you need \r\n \r\n(Yesterday) \r\n(Oh yeah) \r\n(She love you, yeah, yeah, yeah) \r\n(She love you, yeah, yeah, yeah) \r\n(Oh, yesterday)\r\n\r\n"


In [29]:
beatles[beatles.song=='Love Me Do'].text

,text
24805,"Love, love me do. \r\nYou know I love you, \r\nI'll always be true, \r\nSo please, love me do. \r\nWhoa, love me do. \r\n \r\nLove, love me do. \r\nYou know I love you, \r\nI'll always be true, \r\nSo please, love me do. \r\nWhoa, love me do. \r\n \r\nSomeone to love, \r\nSomebody new. \r\nSomeone to love, \r\nSomeone like you. \r\n \r\nLove, love me do. \r\nYou know I love you, \r\nI'll always be true, \r\nSo please, love me do. \r\nWhoa, love me do. \r\n \r\nLove, love me do. \r\nYou know I love you, \r\nI'll always be true, \r\nSo please, love me do. \r\nWhoa, love me do. \r\nYeah, love me do. \r\nWhoa, oh, love me do.\r\n\r\n"


In [30]:
beatles[beatles.song=='If I Needed Someone'].text


,text
24768,"If I needed someone to love \r\nYou're the one that I'd be thinking of \r\nIf I needed someone \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone \r\nAh, ah, ah, ah \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone \r\nAh, ah\r\n\r\n"


In [31]:
beatles[beatles.song=='If I Needed Someone To Love'].text

,text
24769,If I needed someone to love \r\nYoure the woman I'd be thinking of \r\nIf I needed someone \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\n \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\n \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone \r\nIf I needed someone to love \r\nYoure the woman I'd be thinking of \r\nIf I needed someone \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\n \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone \r\n \r\nIf I had some more time to spend \r\nThen I guess I'd be with you my friend \r\nIf I needed someone \r\n \r\nHad you come some other day \r\nThen it might not have been like this \r\nBut you see now I'm too much in love \r\n \r\nCarve your number on my wall \r\nAnd maybe you will get a call from me \r\nIf I needed someone\r\n\r\n


In [32]:
beatles[beatles.song=='Have A Banana!'].text

,text
1221,"[Speech] \r\n \r\nBrian Matthew: Is that it? Is that the end? \r\nPaul: Yeah, yeah, that's it. \r\nJohn: Fade, fade! \r\nBrian: Good track. Oh, well, we'll stop there, stop there, stop there. \r\nJohn: What an end! \r\nBrian: Quiet! All right, George. \r\nJohn: Fade! \r\nBrian: Hold it! \r\nGeorge: Oh, thank you. \r\nJohn: Fade, you silly. \r\nBrian: Well, we did. We did that. Oh, no! No! We've done that bit! \r\nJohn: The train comes in now. \r\nBrian: We did that. \r\nJohn: Yeah. \r\nBrian: To pove we weren't playing the record, then, you see. 'Cause,\r\notherwise, there's no point in you being here, is there? Ha, ha, ha! \r\nJohn: Yeah, we did that, 'cause it sounds just like it, don't it? \r\nBrian: Pretty cool lot of fellows, aren't you? Here, Ringo, have a banana,\r\ncatch!\r\n\r\n"


In [33]:
beatles[beatles.song=='Crinsk Dee Night'].text

,text
24713,"[Speech] \r\n \r\nBrian Matthew: The next few minutes, we're in the lap of the gods and the\r\nhands of the Beatles. In my young days, when I was a lad, they used to have\r\nactors in films and now that they-- \r\nPaul: Yes? \r\nJohn: Hey! Listen! \r\nPaul: It's all changed, now, Brian. They're not doing that, no actors. \r\nJohn: It's all changed, now. \r\nBrian: But this is what I wonder. In those days, the actors used to say their\r\nbest bits were left on the cutting room floor. Did you find that? \r\nJohn: No, no, no, those were the good bits in the film. You should have seen\r\nthe rest. \r\nBrian: Yes? \r\nJohn: Rubbish! \r\nBrian: Was it, really? \r\nJohn: Even worse, yes. \r\nBrian: Who was worst? \r\nJohn: Oh, Paul. \r\nBrian: I see. \r\nPaul: I think John was about the worst. \r\nJohn: No, it was you. \r\nPaul: Oh, Ringo was very good. He was. He's a good lad. \r\nBrian: He was. They're saying he's a new Charlie Chaplin. Do you think that's\r\nright? \r\nJohn: He was miming. \r\nPaul: You, too, with Jason. \r\nJohn: Oh, yes, he's an old one. Okay, Ring? \r\nBrian: Beginning to look like that. Now, then-- \r\nRingo: All right, John. Can you hear me? \r\nPaul: Can you hear him? Hello! \r\nBrian: Not really. I hope not. I hope not! \r\nJohn: We brought you the flowers, Ring. \r\nBrian: He's not allowed to talk. \r\nRingo: Eh? \r\nJohn: We brought you the flowers. \r\nRingo: Oh, good. \r\nJohn: And the grapes. \r\nRingo: Oh, I like grapes. \r\nPaul: He likes grapes. \r\nGeorge: We've been to Portugal. \r\nBrian: Guess who's, guess who's top of the pops in Portugal, then. \r\nAll: Who? \r\nBrian: Los Beatles. \r\nJohn: Los Beatles? Great, great laugh. \r\nBrian: I don't suppose you know the title of your film in Portuguese? \r\nJohn: No. Crinsk dee Night? \r\nBrian: Could be. Let's hear the number, shall we? \r\nJohn: Right\r\n\r\n"


Compare the results. Which Vectorizer seems to do a better job?

*It seems that TF-IDF does a better job at finding the two songs that are the most similar. The count vecotrizer mathces songs based on raw word frequency, ignoring thematic context. On the other hand, TF-IDF would use weights for comparing, and would downweight the most common words, giving importance to words that are distinctive to a specific song, and consequentially, producing more meaningful similarity results*

#type your answer here

#### Supplementary Activity

* Use a dataset that consists of text document
* Remove special characters
* Compare the similar document using Count Vector and TF-IDF Vectorizer
* Use the cosine similarity to measure the similarity of text

In [44]:
import re
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from numpy import dot
from numpy.linalg import norm

The dataset to be used for this supplementary activity is the [20 newsgrops text dataset](https://scikit-learn.org/0.19/datasets/twenty_newsgroups.html), which contains 18000 newsgroups posts on 20 topics.

In [78]:
# Load dataset
newsgroups = fetch_20newsgroups()



In [79]:
# Build a DataFrame
df = pd.DataFrame({'text': newsgroups.data,
                   'label': [newsgroups.target_names[t] for t in newsgroups.target]})

df.tail()

,text,label
11309,"From: jim.zisfein@factory.com (Jim Zisfein) \nSubject: Re: Migraines and scans\nDistribution: world\nOrganization: Invention Factory's BBS - New York City, NY - 212-274-8298v.32bis\nReply-To: jim.zisfein@factory.com (Jim Zisfein) \nLines: 37\n\nDN> From: nyeda@cnsvax.uwec.edu (David Nye)\nDN> A neurology\nDN> consultation is cheaper than a scan.\n\nAnd also better, because a neurologist can make a differential\ndiagnosis between migraine, tension-type headache, cluster, benign\nintracranial hypertension, chronic paroxysmal hemicrania, and other\nheadache syndromes that all appear normal on a scan. A neurologist\ncan also recommend a course of treatment that is appropriate to the\ndiagnosis.\n\nDN> >>Also, since many people are convinced they have brain tumors or other\nDN> >>serious pathology, it may be cheaper to just get a CT scan then have\nDN> >>them come into the ER every few weeks.\nDN> And easier than taking the time to reassure the patient, right?\nDN> Personally, I don't think this can ever be justified.\n\nSigh. It may never be justifiable, but I sometimes do it. Even\nafter I try to show thoroughness with a detailed history, neurologic\nexamination, and discussion with the patient about my diagnosis,\nsalted with lots of reassurance, patients still ask ""why can't you\norder a scan, so we can be absolutely sure?"" Aunt Millie often gets\ninto the conversation, as in ""they ignored Aunt Millie's headaches\nfor years"", and then she died of a brain tumor, aneurysm, or\nwhatever. If you can get away without ever ordering imaging for a\npatient with an obviously benign headache syndrome, I'd like to hear\nwhat your magic is.\n\nEvery once in a while I am able to bypass imaging by getting an EEG.\nMind you, I don't think EEG is terribly sensitive for brain tumor,\nbut the patient feels like ""something is being done"" (as if the\nhours I spent talking with and examining the patient were\n""nothing""), the EEG has no ionizing radiation, it's *much* cheaper\nthan CT or MRI, and the EEG brings in some money to my department.\n---\n . SLMR 2.1 . E-mail: jim.zisfein@factory.com (Jim Zisfein)\n \n",sci.med
11310,"From: ebodin@pearl.tufts.edu\nSubject: Screen Death: Mac Plus/512\nLines: 22\nOrganization: Tufts University - Medford, MA\n\nI have a (very old) Mac 512k and a Mac Plus, both of which \nhave the same problem.\n\nTheir screens blank out, sometimes after a minor physical jolt\n(such as inserting a floppy in the internal drive), sometimes \nall by themselves (computer left to itself just goes blank).\n\nI have replaced the wires connecting the logic boards and the \nvideo board, because it seemed at first that jiggling the wires\nmade the screen come back on. This worked for a while, but the\nblanking out has returned.\n\nCan I do anything? Do I need a new power supply? A new CRT?\nA new computer?\n\nThanks for any advice...\n\n--------------------------\nEthan Bodin\nTufts University\nebodin@pearl.tufts.edu\n--------------------------\n",comp.sys.mac.hardware
11311,"From: westes@netcom.com (Will Estes)\nSubject: Mounting CPU Cooler in vertical case\nOrganization: Mail Group\nX-Newsreader: TIN [version 1.1 PL8]\nLines: 13\n\nI just installed a DX2-66 CPU in a clone motherboard, and tried mounting a CPU \ncooler on the chip. After about 1/2 hour, the weight of the cooler was enough \nto dislodge the CPU from its mount. It ended up bending a few pins\non the CPU, but luckily the power was not on yet. I ended up\npressing the CPU deeply into its socket and then putting the CPU\ncooler back on. So far so good.\n\nHave others had this problem? How do you ensure that the weight of\nthe CPU fan and heatsink do not eventually work the CPU out of its\nsocket when mounting the motherboard in a vertical case?\n\n-- \nWill Estes\t\tInternet: westes@netcom.com\n",comp.sys.ibm.pc.hardware
11312,"From: steve@hcrlgw (Steven Collins)\nSubject: Re: Sphere from 4 points?\nOrganization: Central Research Lab. Hitachi, Ltd

In [80]:
df.shape

(11314, 2)

The loaded data was the training set

In [81]:
# 2. Remove special characters
def clean(text):
    text = re.sub(r'\n', ' ', text)          # remove newlines
    text = re.sub(r'[^a-zA-Z\s]', ' ', text) # remove non-alpha chars
    text = re.sub(r'\s+', ' ', text).strip()  # collapse whitespace
    return text

corpus = df['text'].map(clean)
corpus.tail()

,text
11309,From jim zisfein factory com Jim Zisfein Subject Re Migraines and scans Distribution world Organization Invention Factory s BBS New York City NY v bis Reply To jim zisfein factory com Jim Zisfein Lines DN From nyeda cnsvax uwec edu David Nye DN A neurology DN consultation is cheaper than a scan And also better because a neurologist can make a differential diagnosis between migraine tension type headache cluster benign intracranial hypertension chronic paroxysmal hemicrania and other headache syndromes that all appear normal on a scan A neurologist can also recommend a course of treatment that is appropriate to the diagnosis DN Also since many people are convinced they have brain tumors or other DN serious pathology it may be cheaper to just get a CT scan then have DN them come into the ER every few weeks DN And easier than taking the time to reassure the patient right DN Personally I don t think this can ever be justified Sigh It may never be justifiable but I sometimes do it Even after I try to show thoroughness with a detailed history neurologic examination and discussion with the patient about my diagnosis salted with lots of reassurance patients still ask why can t you order a scan so we can be absolutely sure Aunt Millie often gets into the conversation as in they ignored Aunt Millie s headaches for years and then she died of a brain tumor aneurysm or whatever If you can get away without ever ordering imaging for a patient with an obviously benign headache syndrome I d like to hear what your magic is Every once in a while I am able to bypass imaging by getting an EEG Mind you I don t think EEG is terribly sensitive for brain tumor but the patient feels like something is being done as if the hours I spent talking with and examining the patient were nothing the EEG has no ionizing radiation it s much cheaper than CT or MRI and the EEG brings in some money to my department SLMR E mail jim zisfein factory com Jim Zisfein
11310,From ebodin pearl tufts edu Subject Screen Death Mac Plus Lines Organization Tufts University Medford MA I have a very old Mac k and a Mac Plus both of which have the same problem Their screens blank out sometimes after a minor physical jolt such as inserting a floppy in the internal drive sometimes all by themselves computer left to itself just goes blank I have replaced the wires connecting the logic boards and the video board because it seemed at first that jiggling the wires made the screen come back on This worked for a while but the blanking out has returned Can I do anything Do I need a new power supply A new CRT A new computer Thanks for any advice Ethan Bodin Tufts University ebodin pearl tufts edu
11311,From westes netcom com Will Estes Subject Mounting CPU Cooler in vertical case Organization Mail Group X Newsreader TIN version PL Lines I just installed a DX CPU in a clone motherboard and tried mounting a CPU cooler on the chip After about hour the weight of the cooler was enough to dislodge the CPU from its mount It ended up bending a few pins on the CPU but luckily the power was not on yet I ended up pressing the CPU deeply into its socket and then putting the CPU cooler back on So far so good Have others had this problem How do you ensure that the weight of the CPU fan and heatsink do not eventually work the CPU out of its socket when mounting the motherboard in a vertical case Will Estes Internet westes netcom com
11312,From steve hcrlgw Steven Collins Subject Re Sphere from points Organization Central Research Lab Hitachi Ltd Lines Nntp Posting Host hcrlgw In article qkgbuINNs n shelley u washington edu bolson carson u washington edu Edward Bolson writes Boy this will be embarassing if it is trivial or an FAQ Given points non coplanar how does one find the sphere that is center and radius exactly fitting those points I know how to do it for a circle from points but do not immediately see a straightforward way to do it in D I have checked some geometry books Graphics Gems and 

In [82]:
# 3a. Count Vectorizer document-term matrix
cv = CountVectorizer(stop_words='english')
X_cv = cv.fit_transform(corpus).toarray()

dt_cv = pd.DataFrame(X_cv, columns=cv.get_feature_names_out())
dt_cv.index = df['label']
print("Count Vectorizer shape:", dt_cv.shape)
dt_cv.head()

Count Vectorizer shape: (11314, 88701)


,aa,aaa,aaaa,aaaaaaaaaaaa,aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaauuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuugggggggggggggggg,aaaaagggghhhh,aaaarrgghhhh,aaah,aaahh,aaahhhh,...,zzrk,zzt,zztop,zzum,zzy,zzz,zzzoh,zzzz,zzzzzz,zzzzzzt
label,,,,,,,,,,,,,,,,,,,,,
rec.autos,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
comp.sys.mac.hardware,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
comp.sys.mac.hardware,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
comp.graphics,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
sci.space,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [83]:
# 3b. TF-IDF Vectorizer document-term matrix
tfidf = TfidfVectorizer(stop_words='english')
X_tfidf = tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=tfidf.get_feature_names_out())
dt_tfidf.index = df['label']
print("TF-IDF shape:", dt_tfidf.shape)
dt_tfidf.head()

TF-IDF shape: (11314, 88701)


,aa,aaa,aaaa,aaaaaaaaaaaa,aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaauuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuugggggggggggggggg,aaaaagggghhhh,aaaarrgghhhh,aaah,aaahh,aaahhhh,...,zzrk,zzt,zztop,zzum,zzy,zzz,zzzoh,zzzz,zzzzzz,zzzzzzt
label,,,,,,,,,,,,,,,,,,,,,
rec.autos,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
comp.sys.mac.hardware,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
comp.sys.mac.hardware,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
comp.graphics,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
sci.space,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [84]:
# 4. Cosine similarity function
cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

# Compare doc 1 against all others — Count Vectorizer
query = X_cv[1]
cv_sims = [(cosine(query, X_cv[i]), df['label'].iloc[i], corpus.iloc[i][:60])
           for i in range(len(X_cv)) if i != 1]  # skip doc 1 (comparing against itself)

print("=== Count Vectorizer: Top 5 similar to doc 1 ===")
for score, label, snippet in sorted(cv_sims, reverse=True)[:5]:
    print(f"  [{score:.4f}] ({label}) {snippet}...")

=== Count Vectorizer: Top 5 similar to doc 1 ===
  [0.6607] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject SI clock...
  [0.5458] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Quadra C...
  [0.5286] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re iisi ...
  [0.5212] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re iisi ...
  [0.4687] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re Q at ...


In [85]:
# Compare doc 1 against all others — TF-IDF
query_tfidf = X_tfidf[1]
tfidf_sims = [(cosine(query_tfidf, X_tfidf[i]), df['label'].iloc[i], corpus.iloc[i][:60])
              for i in range(len(X_tfidf)) if i != 1]

print("=== TF-IDF: Top 5 similar to doc 1 ===")
for score, label, snippet in sorted(tfidf_sims, reverse=True)[:5]:
    print(f"  [{score:.4f}] ({label}) {snippet}...")

=== TF-IDF: Top 5 similar to doc 1 ===
  [0.6044] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject SI clock...
  [0.4774] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Quadra C...
  [0.4557] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re iisi ...
  [0.4331] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re iisi ...
  [0.3540] (comp.sys.mac.hardware) From guykuo carson u washington edu Guy Kuo Subject Re Q at ...


The results show that both vectorizers ranked the same subset of data that is most similar to the first post/entry in the dataset. However, the similarity scores differ a bit, with TF-IDF outputting lower similarity scores. This difference show the different approaches of the two vectorizers, with the count vectorizer solely basing the similarity based on the number of similar words, while the TF-IDF vectorizer bases on weights.

#### Conclusion

This activity helped me understand how vectorizers work in processing text data into matrices that can be interpreted mathematically by computers. I was also made aware about one of the applications of this process, which is finding the similarity between two text documents.

In the tasks given, I was able to learn preprocessing techniques when dealing with text data, such as removing special characters from text. I think its important to know how preprocessing works in this type of data as it's a bit different than that of the usual structured data.